Add smoke test suite

pabs3 opened the pull request:
Uses mpv, mpv-mpris, playerctl, sound-theme-freedesktop, bash,
dbus-run-session, xvfb, xauth, jo, jq, socat and awk.

Initial tests include checking metadata, pausing a playing mpv, playing a
paused mpv, toggling mpv between playing and pausing, stopping a playing mpv.

Add enough parameters to make it flexible and useful for testing already
installed mpv mpris plugins, which will primarily be useful for distros.

Document the test requirements and parameters in the README.md file.

Should help prevent breakage and ensure mpv compatibility.

This has been adapted from the Debian autopkgtests that I wrote:

https://sources.debian.org/src/mpv-mpris/unstable/debian/tests/

olifre commented on test/env:
```diff
@@ -0,0 +1,32 @@
+exec \
+env \
+MPV_MPRIS_TEST_NAME="$1" \
+XDG_RUNTIME_DIR="$MPV_MPRIS_TEST_TMP" \
```
This fails unless `MPV_MPRIS_TEST_TMP` is set to something, i.e. I get:
```
+ exec env MPV_MPRIS_TEST_NAME=./metadata XDG_RUNTIME_DIR= dbus-run-session -- xvfb-run -a ./metadata
dbus[32]: Unable to set up transient service directory: XDG_RUNTIME_DIR "" not available: No such file or directory
```
in an environment which has `XDG_RUNTIME_DIR` unset (which is not an issue in general).

olifre replied:
Potentially, something like:
```bash
if [ -n "${MPV_MPRIS_TEST_TMP}" ]; then
  export XDG_RUNTIME_DIR="${MPV_MPRIS_TEST_TMP}";
fi
```
could accomodate the use case to allow for `XDG_RUNTIME_DIR` do stay unset if wanted, and still be overridable.

pabs3 replied:
Thanks, will fix that and I think make the variable a bit less generic.

The thinking with setting `XDG_RUNTIME_DIR` is to avoid polluting the system dirs as much as possible.

olifre replied:
> The thinking with setting XDG_RUNTIME_DIR is to avoid polluting the system dirs as much as possible.

That's a good goal. While of course it won't matter in sandboxed test environments, users may also want to run this manually. It might be helpful to also mention the variable in `README.md` to further encourage use of it ;-).

olifre commented on test/setup:
```diff
@@ -0,0 +1,74 @@
+log_prefix="$MPV_MPRIS_TEST_LOG/$test"
+log="$log_prefix.mpv.log"
+file="$MPV_MPRIS_TEST_PLAY"
+tty="$(tty)"
```
I tried to run these tests as package tests on Gentoo, and failed here due to:
```
+++ tty                                                                                                               
++ tty='not a tty'                                                                                                                                                                                                                          
make[1]: *** [Makefile:6: test] Error 1                                                                                                                                                                                                     
```
For better portability of the test suite, it would of course be nice to be able to run without a `tty`. Maybe wrap everything in `script`?

olifre replied:
Alternatively, maybe use:
```
tee /dev/stderr
```
below, which drops the need for a `tty`.

pabs3 replied:
I'm trying to avoid stderr, since [Debian's CI system by default wants stderr to be empty](https://salsa.debian.org/ci-team/autopkgtest/-/raw/master/doc/README.package-tests.rst) (search for stderr) and I would like to abide by that in case later on there is an mpv message on stderr that indicates some breakage. For examples of that avoidance, see in the setup script were I redirect part of the mpv stderr to stdout and where I used `BASH_XTRACEFD` to redirect `set -x` output to stdout.

I tried `tee /dev/stdout` in the Debian CI and I didn't get any of the JSON strings printed somehow. Seems like that is because it resolves to the stdout of the tee process, which goes into the pipe, which is obviously wrong. So I probably need to either resolve `/dev/stdout` to something first, or just write to a file and then cat the file, which seems like the simplest solution. It seems like another option is process substitution, and since we are depending on bash for `BASH_XTRACEFD`, it seems fine to use process substitution for this.

pabs3 replied:
Process substitution ends up making the indentation etc a bit annoying, so I decided to go with just writing files and using input/output redirection. As a bonus the bash trace output is much more readable this way.

olifre replied:
> I'm trying to avoid stderr, since [Debian's CI system by default wants stderr to be empty](https://salsa.debian.org/ci-team/autopkgtest/-/raw/master/doc/README.package-tests.rst) (search for stderr) and I would like to abide by that in case later on there is an mpv message on stderr that indicates some breakage. 

Indeed, that's a good point (and I also learnt something new about Debian's `autopkgtest`s this way ;-) ). 

The solution you chose sounds great, process substitution is really a bit of a mess to use, Another idea could be named file descriptors, but there's not much of a difference.

pabs3 replied:
Thanks for the named file descriptors tip! That lets me get bash to allocate the `set -x` output file descriptor instead of choosing one and hoping it isn't in use :grinning:

olifre commented:
After changing the parts I commented on, I can confirm the tests run fine in case included in Gentoo package tests, too. Great work :+1: .

pabs3 commented:
Another review would be good, I've made significant improvements to the pull request, including making it check for errors on stderr for each of the tests and making the tests run in parallel (with [GNU make output grouping](https://www.gnu.org/software/make/manual/html_node/Parallel-Output.html)).

olifre commented on test/Makefile:
```diff
@@ -0,0 +1,59 @@
+	tee "$$stderr" ; \
+	< "$$exit_code" read ret ; \
+	if [ -s "$$stderr" ] ; then \
+	  if [ -n "$$MPV_MPRIS_TEST_NO_STDERR" ] ; then
```
This seems to be missing the trailing `\`.

pabs3 replied:
Fixed.

olifre commented:
@pabs3 That looks great! I currently get empty `ipc.output.json` in all cases for some reason (then of course leading to all tests failing), I will investigate this more in-depth later tonight.

olifre commented on test/setup:
```diff
@@ -0,0 +1,96 @@
+	rm -f "$socat_log"
+	# Pass the input JSON to socat and save the output JSON, also redirect
+	# socat errors to a file so we can check afterward if there are errors.
+	socat -lf"$socat_log" "$input_json" "$ipc" > "$output_json"
```
This seems to be `socat` version dependent: Newer versions seem to assume bidirectional communication, even in case it is not useful (as here, when an input file is used). 
For me, using:
```
cat "$input_json" | socat -lf"$socat_log" - "$ipc" > "$output_json"
```
instead works fine.

pabs3 replied:
Actually my socat version also does what yours does, but the tests still passed because my jq expression was fine with an empty output JSON file. To get socat to do all the work itself, you need to use `OPEN:input` and `CREATE:output`, but those are more verbose than - which is a shortcut for STDIO, which is a shortcut for STDIN, STDOUT, which are shortcuts for `FD:0` and `FD:1`, or something, I can't get `OPEN`+`CREATE` to work. I'm going back to what you suggested, adding size checks for the output JSON and will try to figure out how to fix the jq expression.

pabs3 replied:
Found a workaround for the jq issue here: https://github.com/stedolan/jq/issues/1142

olifre commented on README.md:
```diff
@@ -47,6 +47,43 @@ Build requirements:
+   Use a file at most 10 seconds long or the tests will take a long time.
+ - `MPV_MPRIS_TEST_MPV_IPC`: dir for mpv IPC, default is test dir.
+ - `MPV_MPRIS_TEST_DBUS`: dir for D-Bus sockets, default is test dir.
+   Sets the `XDG_RUNTIME_DIR` env var so D-Bus users the dir.
```
`users` => `uses`

pabs3 replied:
Fixed.

olifre reviewed (commented):
This looks great, apart from the version strangeness with `socat` and the missing `\` in one place and a small tpyo I did not find any issues :+1: . 
Since I'm also only a contributor, of course it's up to @hoyon to perform a real review and merge. I think some small extension (in a later commit) could be to also run `make test` (after installing all dependencies) in the GitHub action workflow so the tests run after every commit / PR.

pabs3 commented:
Fixed a few more issues. I think it is ready for a final review before merging.

At this time I don't know enough about GitHub actions to integrate the tests with that, but that can happen at a later stage.

hoyon commented:
Thanks so much for this @pabs3!!!

I haven't been able to review this properly yet but I've checked out the branch and run `make test` and get this error:

```
+ exec env MPV_MPRIS_TEST_NAME=./play XDG_RUNTIME_DIR=. TEMPDIR=. TMPDIR=. TEMP=. TMP=. dbus-run-session -- xvfb-run --auto-servernum --auth-file ././play.Xauthority ./play
xvfb-run: unrecognized option '--auth-file'
test play: stderr not empty!
test play: start of stderr ---------------------
xvfb-run: unrecognized option '--auth-file'
test play: end of stderr -----------------------
test play: stderr not empty!
test play: failed with exit code 1
make[1]: *** [Makefile:18: play] Error 1
make[1]: Leaving directory '/home/hoyon/code/mpv-mpris/test'
make: *** [Makefile:56: test] Error 2
```

Am I missing something? I'm using arch linux and `man xvfb-run` does mention the `--auth-file` flag.

> At this time I don't know enough about GitHub actions to integrate the tests with that, but that can happen at a later stage.

I'd definitely want that. IMO there's not much value in a test suite if it's not being run automatically on every commit.

olifre commented:
> Am I missing something? I'm using arch linux and `man xvfb-run` does mention the `--auth-file` flag.

It seems the flag is listed on the manpage provided for version 21.1.3-2 in Arch here: https://man.archlinux.org/man/xvfb-run.1.en 
Which version are you using?

hoyon commented:
> It seems the flag is listed on the manpage provided for version 21.1.3-2 in Arch here: [man.archlinux.org/man/xvfb-run.1.en](https://man.archlinux.org/man/xvfb-run.1.en) Which version are you using?

It seems like the arch linux xvfb-run script has a number of issues. I've managed to manually patch the script to get the tests to pass but will need to find some time to investigate deeper then report upstream.

pabs3 commented:
@hoyon if I switch from `--auth-file` to `-f`, will that work on Arch Linux?

I added a few more fixes:

 * added the patch from @olifre and fixed it to install `mpv` too
 * moved `--output-sync=target` to `MAKEFLAGS` in `test/Makefile`
 * moved the contents of the tests targets in `test/Makefile` to a `test/wrapper` script, tons of shell in `Makefile`s looks ridiculously ugly to me
 * switched to a better jq command-line, recommended to me on their IRC channel
 * fixed a shellcheck warning in the `wrapper` script
 * added a GitHub check that the build process doesn't modify files, that `.gitignore` is correct and `make clean` is correct.
 * fixed the README to reference dbus instead of dbus-daemon, since that only appears in Debian bookworm and later, while just `dbus` is available in all releases and pulls in dbus-daemon in bullseye.

pabs3 commented:
One more fix: avoid ././ in the Xauthority path.

-- 
bye,
pabs

https://bonedaddy.net/pabs3/

hoyon commented on test/setup:
```diff
@@ -0,0 +1,96 @@
+
+
+prop () {
+	prop="$1"
+	shift
+	jo command[]=get_property command[]="$prop" "$@"
```
Can we use jq instead of jo to remove one of the test dependencies here?

```suggestion
	jq -n -c --arg prop "$1" '{command: ["get_property", $prop]}'

```

pabs3 replied:
Fixed.

hoyon commented on test/env:
```diff
@@ -0,0 +1,87 @@
+TEMP="$MPV_MPRIS_TEST_TMP" \
+TMP="$MPV_MPRIS_TEST_TMP" \
+dbus-run-session -- \
+xvfb-run --auto-servernum --auth-file "$MPV_MPRIS_TEST_XAUTH/$(basename "$1").Xauthority" \
```
Replacing this with -f does indeed work in arch linux. I've reported the bug upstream.

pabs3 replied:
Fixed, added a comment about it.

What is the bug link?

hoyon replied:
https://bugs.archlinux.org/task/73865

pabs3 replied:
Added to the patch. Looks like it is fixed now so maybe I can switch back to the long option though?

hoyon commented on test/Makefile:
```diff
@@ -0,0 +1,28 @@
+	test \
+	$(tests) \
+	clean
+
```
The tests fail if running make with parallel jobs. Took me a while to realise the tests were failing on my machine because I have a terminal alias for make=make -j24

The easy solution would be to add a [.NOTPARALLEL](https://www.gnu.org/software/make/manual/make.html#index-_002eNOTPARALLEL) directive for these test jobs or fix running them in parallel.

pabs3 replied:
The tests work in parallel for me, how are you running them? I use this:

```
make -j10 test
```

Each test should be using independent files/dirs/sockets so I'm surprised they break in parallel for you.

hoyon replied:
running `make -j10 test` on my machine gives errors like these at the end of some tests:

```
/usr/bin/xvfb-run: line 186: kill: (5825) - No such process
test play-pause: stderr not empty!
test play-pause: start of stderr ---------------------
/usr/bin/xvfb-run: line 186: kill: (5825) - No such process
test play-pause: end of stderr -----------------------
test play-pause: stderr not empty!
test play-pause: failed with exit code 1
make[1]: *** [Makefile:18: play-pause] Error 1
make[1]: Leaving directory '/home/hoyon/code/mpv-mpris/test'
make: *** [Makefile:56: test] Error 2
```

Possibly another bug in the arch linux xvfb-run.

pabs3 replied:
Thanks, I'll try to download a copy of their xvfb-run and then run it on Debian.

pabs3 replied:
BTW, can you try a parallel test with the new version that fixed the bug you linked above?

pabs3 replied:
OK, I reproduced the failure with their version.

pabs3 replied:
OK, switching it to use `--auto-display` instead of `--auto-servernum` makes it work. Unfortunately the Debian `xvfb-run` does not have `--auto-display` though, but the `--auto-servernum` works fine with it. So I think that the right approach is to grep the `--help` output and use `--auto-display` when it is present.

pabs3 replied:
Implemented that, fixes the parallel failure and works on Debian too.

hoyon reviewed (changes requested):
Apologies for the delay!

Some minor comments but thanks again for this!

pabs3 commented:
Updated to add an xvfb log too.

hoyon reviewed (approved):
Tests all pass when running in parallel now!

Thanks again for this work. I'll merge it in now!

pabs3 commented:
Thanks for merging.

Any particular reason you do squashes instead of merges or rebases?

Can we get a new release for the cleanups and test suite?

-- 
bye,
pabs

https://bonedaddy.net/pabs3/
